grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -375,6 +375,7 @@ class StockMovementService {
         def offset = params.offset ? params.int("offset") : null
         Date createdAfter = params.createdAfter ? Date.parse("MM/dd/yyyy", params.createdAfter) : null
         Date createdBefore = params.createdBefore ? Date.parse("MM/dd/yyyy", params.createdBefore) : null
+        List<ShipmentType> shipmentTypes = params.list("shipmentType")?.collect{ ShipmentType.read(it) } ?: null
```

Comment:
I don't mind this, just make sure that we don't get 2N queries for this. In other words, for the normal way to do this would be to check whether the parameter exists and then do the lookup.
```
params.list("shipmentType") ? params.list("shipmentType").collect{ ShipmentType.read(it) } : null
```
whereas your code might be doing the same queries twice once to check if the list is null and again as part of the return of the elvis operator. 

I've never understood what happens here (whether it evaluates again or caches and returns the value from the null check).

Reply:
I honestly do not understand the point or I missed something.
From my understanding and knowledge how it works, if I do this:
```groovy
params.list("shipmentType")?.collect{ ShipmentType.read(it) } ?: null
```
The `.collect` doesn't get evaluted if the `params.list("shipmentType")` is `null` and "goes further" if it is not `null`.

So summing up, in my opinion both - my version and your version do the same thing.
But I love to hear from you a further explanation/thoughts, because I've probably missed something, but this was my understanding of those expressions.

Reply:
I am just wondering if under the covers the elvis operator (`?:`) 
```
params.list("shipmentType").collect{ ShipmentType.read(it) } ?: null
```
does this
```
def shipmentTypes = params.list("shipmentType").collect{ ShipmentType.read(it) }
if shipmentTypes != null 
    return params.list("shipmentType").collect{ ShipmentType.read(it) }
else 
    return null
```
or this 
```
def shipmentTypes = params.list("shipmentType").collect{ ShipmentType.read(it) }
if shipmentTypes != null 
    return shipmentTypes
else 
    return null
```

The former would likely cause [negligible] performance degradation since it's doing all of those database reads twice. Obviously, we can probably live with that, but I don't want to get in the habit of doing that unless we know which it is.

Reply:
It does neither I believe.
What happens behind this: `params.list("shipmentType")?.collect{ ShipmentType.read(it) }` is that:
```groovy
def shipmentType = params.list("shipmentType")
if (shipmentType != null) {
  return shipmentType.collect { ShipmentType.read(it) }
}
return null
```
so the `.collect.` doesn't get evaluated if `params.list("shipmentType")` is `null`, so the first scenario you've given cannot happen, because null safe operator prevents us from doing an operation if something before it is `null`, so the `.collect` can't be evaluted it the `params.list("shipmentType")` is `null`.

> "In Groovy we can do this shorthand by using the null safe operator (?.). If the variable before the question mark is null it will not proceed and actually returns null for you."

So, again, in my opinion both: my version and this version:
```groovy
params.list("shipmentType") ? params.list("shipmentType").collect{ ShipmentType.read(it) } : null
```
do the same thing, but I believe my version does it even more "efficiently", since in worst case scenario the above might call `params.list("shipmentType")` twice, wheras my version should just use "one" reference to `params.list("shipmentType")`.


Now I even actually believe, that I could just leave this:
```groovy
params.list("shipmentType")?.collect{ ShipmentType.read(it) }
```
without having to do the elvis operator.

Reply:
After a quick check, I need to agree with your version - it isn't because of some 2N queries, because all I've explained is right, so I hope everyone could have some value from it, but basically why I brought back your version has a different reason:
```groovy
params.list("shipmentTypes")
```
even if there is no `params.shipmentTypes`, it evaluates to an empty list, so with my version:
```groovy
params.list("shipmentType")?.collect{ ShipmentType.read(it) } 
```
I was doing unnecessary collect on empty list, because `?` only checks if it is `null`, not if the value is truthy. I just thought `params.list` doesn't evalute do `[]`, but to `null`.

Anyway, thanks for that discussion, as there was an occasion to double check some things and consolidate some knowledge
```